<html>

<head>
<LINK REL="Stylesheet" TYPE="text/css" HREF="stylesheet.css">
<title>PCR and Primer Design</title>
</head>

<body>

<h1>PCR and Primer Design</h1>
The PCR functionalityis invoked through "Tools/PCR" or the context menu of the plasmid map.
The latter offers, if a sequence is selected, the generation of "forward", "backward", or "both" primers.
<p>The PCR display consists of a list of primers, a primer detail view, and a large sequence part,
which in arranged in block of seven lines:</p>

<ol>
<li>5' primer</li>
<li>Original DNA</li>
<li>Original amino acid sequence</li>
<li>Original DNA (complementary)</li>
<li>3' primer</li>
<li>Restriction sites of the</li>
<li>resulting DNA sequence</li>
<li>Resulting amino acid sequence</li>
</ol>

<center><img src="de_pcr.png"></center>

<p>The amino acid sequences #3 and #7 have the same reading frame.</p>
<p>Primers can be edited through "View/Edit mode" or "F2". Selecting the sequence
of the 5'- or the 3'-primer prior to editing chooses which primer is edited.</p>
<p>Primer can be imported and exported through buttons in the toolbar.
Primers to be imported have to be loaded from the database first.
Exported primers have to be manually stored in the databese.</p>
<p>Primer can be optimized semi-automatically. The "Edit" button
opens a dialog for the primer selected in the list.
According to the settings in theis dialog, a variant of the primer can be
chosen by double-clicking on it in the list.</p>
<p>A silent mutation can be found in a dialog invokable through the context menu.
A sequence (usually a primer) has to be selected, and a reading frame has to be
chosen for the silent mutation function to be available.</p>

<center><a href="en_main.htm">Main page</a></center>
</body>

</html>